File: QueueUsingLinkedList.java
package l14;


 class Nodes<T> {
	public T data;
	public Nodes<T> next;

	public Nodes(T data) {
		this.data = data;
		this.next = null;
	}
}

public class QueueUsingLinkedList<T> {
	private Nodes<T> front;
	private int length;
	private Nodes<T> back;

	public QueueUsingLinkedList() {
		front = null;
		length = 0;
		back = front;
	}

	public int size() {
		return length;
	}

	public T frontElement() throws QueueEmptyException {
		if (size()==0) {
			throw new QueueEmptyException("");
		}
		return front.data;
	}

	public void enqueue(T element) {
		Nodes<T> newNode = new Nodes<T>(element);
		if (back != null)  {
			back.next = newNode;
			back = newNode;
		} else {
			front = newNode;
			back = newNode;
		}

		length++;
	}

	public boolean isEmpty(){
		if (size()==0) {
			return true;
		}
		else
			return false;
	}

	public T dequeue() throws QueueEmptyException {

		if (size()==0) {
			throw new QueueEmptyException(" ");
		}
		T dataToBeRemoved = front.data;
		front = front.next;
		length--;
		if (front == null) {
			back = null;
		}
		return dataToBeRemoved;
	}

}

File: QueueUsingLinkedList.cpp
#include <iostream>
using namespace std;

struct node{
    node *next;
    int data;
};

class myQueue
{
    node *frontN,*rearN;

public:
    myQueue(){
        frontN = NULL;
        rearN = NULL;
    }

    void enqueue(int value){
        node *temp = new node;
        temp -> data = value;

        if(frontN == NULL  ){
            frontN = temp;
            rearN = temp;
            frontN -> next = NULL;
        }else{
            rearN -> next = temp;
            temp -> next = NULL;
            rearN = temp;
        }
        cout<<value<<" enqueued \n";
    }

    void dequeue(){
        if(frontN == NULL){
            cout<<"Already Empty Queue \n";
        }else if(frontN == rearN){
            cout<<frontN -> data<<" dequeued , now empty Queue \n";
             frontN = NULL;
             rearN = NULL;
        }else{
            cout<<frontN -> data<<" dequeued \n";
            frontN  = frontN -> next;
        }
    }

    void display(){
        node *temp = frontN;
        while(temp != rearN -> next){
                cout<<temp -> data<<"\t";
                temp = temp -> next;
        }
        cout<<endl;
    }

    void isEmpty(){
        if(frontN == NULL){
            cout<<"Queue is empty \n";
        }else{
            cout<<"Queue is not empty \n";
        }
    }
};
int main()
{
    myQueue q1;
    q1.enqueue(1);
    q1.display();
    q1.enqueue(10);
    q1.display();
    q1.enqueue(20);
    q1.display();
    q1.dequeue();
    q1.display();
    q1.isEmpty();
    return 0;
}
